arangod/RocksDBEngine/RocksDBSha256Checksum.cpp

```diff
@@ -0,0 +1,212 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBSha256Checksum.h"
+#include "Basics/Exceptions.h"
+#include "Basics/StringUtils.h"
+#include "Basics/debugging.h"
+#include "Basics/files.h"
+#include "Basics/FileUtils.h"
+#include "Basics/MutexLocker.h"
+
+namespace arangodb {
+
+RocksDBSha256Checksum::RocksDBSha256Checksum(std::string const& filename, std::shared_ptr<RocksDBShaFileManager> shaFileManager)
+    : _fileName(filename),
+      _shaFileManager{shaFileManager},
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+      _context(EVP_MD_CTX_new()) {
+#else
+      _context(EVP_MD_CTX_create()) {
+#endif
+  if (_context == nullptr) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+  }
+  if (EVP_DigestInit_ex(_context, EVP_sha256(), nullptr) == 0) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "unable to initialize SHA256 processor");
+  }
+}
+
+RocksDBSha256Checksum::~RocksDBSha256Checksum() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void RocksDBSha256Checksum::Update(char const* data, size_t n) {
+  EVP_DigestUpdate(_context, static_cast<void const*>(data), n);
+}
+
+void RocksDBSha256Checksum::Finalize() {
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int lengthOfHash = 0;
+  if (EVP_DigestFinal_ex(_context, hash, &lengthOfHash) == 0) {
+    TRI_ASSERT(false);
+  }
+  _checksum = basics::StringUtils::encodeHex(reinterpret_cast<char const*>(&hash[0]), lengthOfHash);
+
+  _shaFileManager->storeShaItems(_fileName, _checksum);
+}
+
+std::string RocksDBSha256Checksum::GetChecksum() const {
+  TRI_ASSERT(!_checksum.empty());
+  return _checksum;
+}
+  
+std::unique_ptr<rocksdb::FileChecksumGenerator> RocksDBSha256ChecksumFactory::CreateFileChecksumGenerator(
+    rocksdb::FileChecksumGenContext const& context) {
+  if (context.requested_checksum_func_name == "RocksDBSha256Checksum") {
+    return std::make_unique<RocksDBSha256Checksum>(context.file_name, _shaFileManager);
+  }
+  return std::make_unique<RocksDBSha256Checksum>(context.file_name, _shaFileManager);
+  // for everything else fall back to CRC32 checksum
+ // LOG_DEVEL << (void*) this <<  " GENERATING CRC";
+ // return std::make_unique<rocksdb::FileChecksumGenCrc32c>(context);
```

Comment:
Do we need the crc32 checksum for anything still? I am not 100% sure that we don't need it for some older files from older versions. We should try this out.
Code for `if` and `else` are
```